Overloading connect with factory

tgriesser opened the pull request:
I had actually considered this before doing the `connectFactory` approach in #185. The issue was that it felt wrong to overload the arguments and require first calling the function to determine its return type.

That said, this is the cleanest implementation I could come up with. It checks if the argument in question has `fn.length` of 0 and also returns a function.

This breaks a few tests, but nothing critical - only on some spies or guarantees an arity-0 function is called a certain amount of times. I can clean those up if this looks alright otherwise.

gaearon commented:
Rather than look at arg count, can we determine whether they returned functions instead of objects lazily first time we call them from the constructor?

tgriesser commented:
Took another pass at it, it's a little more verbose / complicated, but does things lazily as you mentioned, all tests are passing and it avoids calling the mapping functions any more times than necessary. 

If this looks alright I can add some tests.

gaearon commented on src/components/connect.js:
```diff
@@ -14,65 +14,40 @@ const defaultMergeProps = (stateProps, dispatchProps, parentProps) => ({
   ...dispatchProps
 })
 
+function isFunction(fn) {
```
I would leave those checks inline.

gaearon commented:
Yes, this looks good. Please do add tests.

tgriesser commented:
Updated, rebased, tests added. Also noticed when writing the tests that a final shallow equality check on the final merged props would save a render if there is a custom `mergeProps` supplied and nothing has changed, so I added that as well.

gaearon commented on test/components/connect.spec.js:
```diff
@@ -1509,5 +1509,108 @@ describe('React', () => {
+      expect(updatedCount).toBe(0)
+      expect(memoizedReturnCount).toBe(2)
+    })
+
```
Nit: mind killing the newline here?

gaearon commented on src/components/connect.js:
```diff
@@ -18,51 +18,33 @@ function getDisplayName(WrappedComponent) {
   const finalMergeProps = mergeProps || defaultMergeProps
-  const doStatePropsDependOnOwnProps = finalMapStateToProps.length !== 1
-  const doDispatchPropsDependOnOwnProps = finalMapDispatchToProps.length !== 1
+  const checkMergedEquals = finalMergeProps !== defaultMergeProps
```
Can we add a new test for this optimization so we don't regress?

gaearon commented on src/components/connect.js:
```diff
@@ -96,8 +78,47 @@ export default function connect(mapStateToProps, mapDispatchToProps, mergeProps,
+      computeDispatchProps(store, props) {
+        if (!this.finalMapDispatchToProps) {
+          return this.configureFinalMapDispatch(store, props)
+        }
```
Nit: please add a newline here

gaearon commented on src/components/connect.js:
```diff
@@ -96,8 +78,47 @@ export default function connect(mapStateToProps, mapDispatchToProps, mergeProps,
+        const dispatchProps = this.doDispatchPropsDependOnOwnProps ?
+          this.finalMapDispatchToProps(dispatch, props) :
+          this.finalMapDispatchToProps(dispatch)
+        return checkStateShape(dispatchProps, true)
```
Nit: please, in general, format more sparsely.

```js
if (!this.finalMapDispatchToProps) {
  return this.configureFinalMapDispatch(store, props)
}

const { dispatch } = store
const dispatchProps = this.doDispatchPropsDependOnOwnProps ?
  this.finalMapDispatchToProps(dispatch, props) :
  this.finalMapDispatchToProps(dispatch)

return checkStateShape(dispatchProps, true)
```

gaearon commented on src/components/connect.js:
```diff
@@ -96,8 +78,47 @@ export default function connect(mapStateToProps, mapDispatchToProps, mergeProps,
+        const isFactory = typeof mappedDispatch === 'function'
+        this.finalMapDispatchToProps = isFactory ? mappedDispatch : mapDispatch
+        this.doDispatchPropsDependOnOwnProps = this.finalMapDispatchToProps.length !== 1
+        return isFactory ? this.computeDispatchProps(store, props) : checkStateShape(mappedDispatch, true)
```
Nit: Would be great to break this ternary into three lines like the rest of them

gaearon commented:
That’s amazing, thank you so much for your work on this. A few more nits and it’s good to go.

tgriesser commented:
All set with a test and a bit of formatting polish

gaearon commented:
Would you be up to contributing a section to [Computing Derived Data](https://github.com/rackt/redux/blob/master/docs/recipes/ComputingDerivedData.md) recipe documenting that? I think a lot of people won’t learn about this useful feature otherwise.

tgriesser commented:
Sure thing, I'll take a look at doing this a little later.

ellbee commented:
@tgriesser I have started to update the Reselect docs to reflect this change. If you haven't started on this already, would you like me to carry on and document it?
